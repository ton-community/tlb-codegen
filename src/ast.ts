export class TLBBinaryOp {
  constructor(
    readonly left: TLBMathExpr,
    readonly right: TLBMathExpr,
    readonly operation: string,
    readonly variables: Set<string> = new Set([
      ...left.variables,
      ...right.variables,
    ]),
    readonly hasNeg: boolean = left.hasNeg || right.hasNeg
  ) {}
}

export class TLBNumberExpr {
  constructor(
    readonly n: number,
    readonly variables: Set<string> = new Set<string>(),
    readonly hasNeg: boolean = false
  ) {}
}

export class TLBUnaryOp {
  constructor(
    readonly value: TLBMathExpr,
    readonly operation: string,
    readonly variables: Set<string> = value.variables,
    readonly hasNeg: boolean = value.hasNeg
  ) {}
}

export class TLBVarExpr {
  constructor(
    readonly x: string,
    readonly variables: Set<string> = new Set<string>(),
    readonly hasNeg: boolean = false
  ) {
    if (variables.size == 0) {
      variables.add(x);
    }
  }
}

export type TLBMathExpr = TLBBinaryOp | TLBNumberExpr | TLBVarExpr | TLBUnaryOp;

export type TLBVariableType = "Type" | "#";

export class TLBVariable {
  constructor(
    readonly isConst: boolean,
    readonly negated: boolean,
    readonly type: TLBVariableType,
    readonly name: string,
    readonly isField: boolean,
    readonly deriveExpr?: TLBMathExpr,
    readonly initialExpr?: TLBMathExpr
  ) {}
}

export type TLBBitsType = {
  kind: "TLBBitsType";
  bits: TLBMathExpr;
};

export type TLBNamedType = {
  kind: "TLBNamedType";
  name: string;
  arguments: TLBFieldType[];
};

export type TLBMathExprType = {
  kind: "TLBExprMathType";
  expr: TLBMathExpr;
};

export type TLBBoolType = {
  kind: "TLBBoolType";
};

export type TLBCoinsType = {
  kind: "TLBCoinsType";
}

export type TLBAddressType = {
  kind: "TLBAddressType";
  addrType: "Internal" | "External" | "Any"
};

export type TLBVarIntegerType = {
  kind: "TLBVarIntegerType";
  signed: boolean;
  n: TLBMathExpr;
}

export type TLBHashmapType = {
  kind: "TLBHashmapType";
  key: TLBMathExprType,
  value: TLBFieldType
}

export type TLBCellType = {
  kind: "TLBCellType";
};

export type TLBNumberType = {
  kind: "TLBNumberType";
  bits: TLBMathExpr;
  signed: boolean;
  storeBits: TLBMathExpr;
  maxBits: number | undefined;
};

export type TLBNegatedType = {
  kind: "TLBNegatedType";
  variableName: string;
};

export type TLBCellInsideType = {
  kind: "TLBCellInsideType";
  value: TLBFieldType;
};

export type TLBMultipleType = {
  kind: "TLBMultipleType";
  value: TLBFieldType;
  times: TLBMathExpr;
};

export type TLBCondType = {
  kind: "TLBCondType";
  value: TLBFieldType;
  condition: TLBMathExpr;
};

export type TLBExoticType = {
  kind: "TLBExoticType";
};

export type TLBFieldType =
  | TLBNumberType
  | TLBBitsType
  | TLBNamedType
  | TLBBoolType
  | TLBCoinsType
  | TLBAddressType
  | TLBHashmapType
  | TLBVarIntegerType
  | TLBCellType
  | TLBMathExprType
  | TLBNegatedType
  | TLBCellInsideType
  | TLBMultipleType
  | TLBCondType
  | TLBExoticType;

export type TLBField = {
  name: string;
  anonymous: boolean;
  fieldType: TLBFieldType;
  subFields: TLBField[];
};

export class TLBParameter {
  constructor(
    readonly variable: TLBVariable,
    readonly paramExpr: TLBMathExpr,
    readonly argName?: string
  ) {}
}

export type TLBConstructorTag = {
  bitLen: number;
  binary: string;
};

export class TLBConstructor {
  constructor(
    readonly parameters: Array<TLBParameter>,
    readonly variables: Array<TLBVariable>,
    readonly variablesMap: Map<string, TLBVariable>,
    readonly parametersMap: Map<string, TLBParameter>,
    readonly name: string,
    readonly fields: Array<TLBField>,
    readonly tag: TLBConstructorTag,
    readonly constraints: Array<TLBMathExpr>,
    readonly declaration: string,
    readonly tlbType: string
  ) {}
}

export class TLBType {
  constructor(
    readonly name: string,
    readonly constructors: Array<TLBConstructor>
  ) {}
}

export class TLBCode {
  constructor(readonly types: Map<string, TLBType>) {}
}
