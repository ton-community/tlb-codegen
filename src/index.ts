export * from './ast';
export { getTLBCodeByAST, generateCodeByAST, getTLBCode, generateCodeWithGenerator, generateCode } from './main';
export { isBigInt, isBigIntExpr } from './generators/typescript/utils';
export type { CodeGenerator } from './generators/generator';
export { TypescriptGenerator } from './generators/typescript/generator';
