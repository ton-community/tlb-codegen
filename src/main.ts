import { Program } from "@igorivaniuk/tlb-parser/dist/ast/nodes";
import { TLBCode, TLBType } from "./ast";
import {
  TLBCodeBuild,
  TLBTypeBuild,
} from "./astbuilder/utils";
import { convertCodeToReadonly } from "./astbuilder/fill_constructors";
import { fillConstructors } from "./astbuilder/fill_constructors";
import { CodeBuilder } from "./generators/CodeBuilder";
import { CodeGenerator, CommonGenDeclaration } from "./generators/generator";
import { TypescriptGenerator } from "./generators/typescript/generator";

export function generate(tree: Program, input: string) {
  let oldTlbCode: TLBCodeBuild = { types: new Map<string, TLBTypeBuild>() };

  let splittedInput = input.split("\n");

  fillConstructors(tree.declarations, oldTlbCode, splittedInput);
  let tlbCode: TLBCode = convertCodeToReadonly(oldTlbCode);

  let codeGenerator: CodeGenerator = new TypescriptGenerator(tlbCode);

  codeGenerator.addTonCoreClassUsage("Builder");
  codeGenerator.addTonCoreClassUsage("Slice");
  codeGenerator.addTonCoreClassUsage("beginCell");
  codeGenerator.addTonCoreClassUsage("BitString");
  codeGenerator.addTonCoreClassUsage("Cell");
  codeGenerator.addTonCoreClassUsage("Address");
  codeGenerator.addTonCoreClassUsage("ExternalAddress");
  codeGenerator.addTonCoreClassUsage("Dictionary")
  codeGenerator.addTonCoreClassUsage("DictionaryValue")

  codeGenerator.addBitLenFunction();

  let jsCodeDeclarations: CommonGenDeclaration[] = [];
  codeGenerator.jsCodeDeclarations.forEach((declaration) => {
    jsCodeDeclarations.push(declaration);
  });

  tlbCode.types.forEach((tlbType: TLBType) => {
    codeGenerator.addTlbType(tlbType);
  });

  let generatedCode = "";

  codeGenerator.jsCodeConstructorDeclarations.forEach((element) => {
    jsCodeDeclarations.push(element);
  });

  codeGenerator.jsCodeFunctionsDeclarations.forEach((element) => {
    jsCodeDeclarations.push(element);
  });

  jsCodeDeclarations.forEach((element) => {
    generatedCode +=
      codeGenerator.toCode(element, new CodeBuilder()).render() + "\n";
  });

  return generatedCode;
}
