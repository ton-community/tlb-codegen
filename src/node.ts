import fs from 'fs/promises';

import { ast } from '@ton-community/tlb-parser';

import { TLBCode } from './ast';
import { CodeGenerator } from './generators/generator';
import { generateCodeByAST, getGenerator, getTLBCodeByAST } from './main';

export async function getTLBCode(inputPath: string) {
    const input = await fs.readFile(inputPath, 'utf-8');

    const tree = ast(input);

    return getTLBCodeByAST(tree, input);
}

export async function generateCodeWithGenerator(
    inputPath: string,
    outputPath: string,
    getGenerator: (tlbCode: TLBCode) => CodeGenerator,
) {
    const input = await fs.readFile(inputPath, 'utf-8');

    const tree = ast(input);

    await fs.writeFile(outputPath, generateCodeByAST(tree, input, getGenerator), {});
    // eslint-disable-next-line no-console
    console.log(`Generated code is saved to ${outputPath}`);
}

export async function generateCode(inputPath: string, outputPath: string, resultLanguage: string) {
    return generateCodeWithGenerator(inputPath, outputPath, getGenerator(resultLanguage));
}
