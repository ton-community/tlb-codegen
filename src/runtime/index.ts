import { ast, Program } from "@ton-community/tlb-parser";
import {
  TLBCode,
  TLBType,
  TLBConstructor,
  TLBField,
  TLBFieldType,
  TLBMathExpr,
  TLBNumberExpr,
  TLBVarExpr,
  TLBBinaryOp,
  TLBParameter,
} from "../ast";
import { getTLBCodeByAST } from "../main";
import {
  Builder,
  Slice,
  Cell,
  beginCell,
  BitString,
  Address,
  ExternalAddress,
  Dictionary,
  parseTuple,
  serializeTuple,
} from "@ton/core";

// Interpreter to evaluate TLBMathExpr at runtime
export class MathExprEvaluator {
  private variables: Map<string, number>;

  constructor(variables: Map<string, number> = new Map()) {
    this.variables = variables;
  }

  evaluate(expr: TLBMathExpr): number {
    if (expr instanceof TLBNumberExpr) {
      return expr.n;
    }
    if (expr instanceof TLBVarExpr) {
      const value = this.variables.get(expr.x);
      if (value === undefined) {
        throw new Error(`Variable ${expr.x} is not defined`);
      }
      return value;
    }
    if (expr instanceof TLBBinaryOp) {
      const left = this.evaluate(expr.left);
      const right = this.evaluate(expr.right);
      switch (expr.operation) {
        case "+":
          return left + right;
        case "-":
          return left - right;
        case "*":
          return left * right;
        case "/":
          return Math.floor(left / right);
        case "%":
          return left % right;
        case "<<":
          return left << right;
        case ">>":
          return left >> right;
        case "&":
          return left & right;
        case "|":
          return left | right;
        case "^":
          return left ^ right;
        case "==":
          return left === right ? 1 : 0;
        case "!=":
          return left !== right ? 1 : 0;
        case "<":
          return left < right ? 1 : 0;
        case "<=":
          return left <= right ? 1 : 0;
        case ">":
          return left > right ? 1 : 0;
        case ">=":
          return left >= right ? 1 : 0;
        default:
          throw new Error(`Unknown operation: ${expr.operation}`);
      }
    }
    // TLBUnaryOp
    const value = this.evaluate((expr as any).value);
    const operation = (expr as any).operation;
    switch (operation) {
      case "-":
        return -value;
      case "~":
        return ~value;
      case "!":
        return value ? 0 : 1;
      default:
        throw new Error(`Unknown unary operation: ${operation}`);
    }
  }
}

// Runtime TL-B serialization/deserialization
export class TLBRuntime {
  private tlbCode: TLBCode;

  constructor(tlbCode: TLBCode) {
    this.tlbCode = tlbCode;
  }

  static fromTLB(tlbSource: string): TLBRuntime {
    const tree = ast(tlbSource);
    const tlbCode = getTLBCodeByAST(tree, tlbSource);
    return new TLBRuntime(tlbCode);
  }

  // Deserialize data from a Slice based on a TL-B type name
  deserialize(typeName: string, slice: Slice): any {
    const type = this.tlbCode.types.get(typeName);
    const types = this.tlbCode.types.keys();
    console.log(
      Array.from(types).filter((p) => p.toLowerCase().includes("transaction"))
    );
    if (!type) {
      throw new Error(`Type ${typeName} not found in TL-B schema`);
    }

    return this.deserializeType(type, slice);
  }

  // Serialize data to a Builder based on a TL-B type name
  serialize(typeName: string, data: any): Builder {
    const type = this.tlbCode.types.get(typeName);
    if (!type) {
      throw new Error(`Type ${typeName} not found in TL-B schema`);
    }

    const builder = beginCell();
    this.serializeType(type, data, builder);
    return builder;
  }

  private deserializeType(
    type: TLBType,
    slice: Slice,
    args: TLBFieldType[] = []
  ): any {
    // Try each constructor in the type
    console.log(type.constructors);
    for (const constructor of type.constructors) {
      try {
        // Save position to restore if the constructor doesn't match
        const bits = slice.remainingBits;
        const refs = slice.remainingRefs;

        // Check tag if present
        if (constructor.tag.bitLen > 0) {
          const preloadedTag = slice.preloadUint(constructor.tag.bitLen);
          const expectedTag = parseInt(constructor.tag.binary.slice(2), 2);
          if (preloadedTag !== expectedTag) {
            // Restore position and try next constructor
            // Reset bit position
            slice.skip(-bits + slice.remainingBits);

            // Reset refs by loading them if needed
            const refsToReset = -refs + slice.remainingRefs;
            for (let i = 0; i < refsToReset; i++) {
              slice.loadRef();
            }
            continue;
          }
          // Consume the tag
          slice.loadUint(constructor.tag.bitLen);
        }

        // Initialize variables map for constraint evaluation
        const variables = new Map<string, number>();

        // Deserialize fields
        let result: any = {
          kind: `${type.name}_${constructor.name}`,
        };

        for (const field of constructor.fields) {
          console.log(
            "result",
            result,
            "remaining bits",
            slice.remainingBits,
            "remaining refs",
            slice.remainingRefs,
            "field",
            field
          );

          // field.subFields.length
          if (field.subFields.length > 0) {
            const ref = slice.loadRef();
            const refSlice = ref.beginParse(true);
            const subfields: any = {};
            for (const subfield of field.subFields) {
              subfields[subfield.name] = this.deserializeField(
                subfield,
                refSlice,
                variables
              );
            }

            if (!field.anonymous) {
              result[field.name] = subfields;
            } else {
              result = { ...result, ...subfields };
            }
          } else {
            if (
              field.fieldType.kind === "TLBNamedType" &&
              constructor.parametersMap.get(field.fieldType.name)
            ) {
              const param = constructor.parametersMap.get(
                field.fieldType.name
              ) as TLBParameter;
              const paramIndex = constructor.parameters.findIndex(
                (p) => p.variable.name === param.variable.name
              );
              field.fieldType = args[paramIndex];
            }

            const fieldValue = this.deserializeField(field, slice, variables);

            if (!field.anonymous) {
              result[field.name] = fieldValue;
            }
          }
        }

        // Check constraints
        const evaluator = new MathExprEvaluator(variables);
        for (const constraint of constructor.constraints) {
          console.log("constraint", constraint);
          if (evaluator.evaluate(constraint) !== 1) {
            // Constraint failed, try next constructor
            // Reset bit position
            slice.skip(-bits + slice.remainingBits);

            // Reset refs by loading them if needed
            const refsToReset = -refs + slice.remainingRefs;
            for (let i = 0; i < refsToReset; i++) {
              slice.loadRef();
            }
            continue;
          }
        }

        return result;
      } catch (e) {
        // If deserialization fails, try the next constructor

        console.error(e);
        continue;
      }
    }

    throw new Error(
      `Failed to deserialize type ${type.name}: no matching constructor found`
    );
  }

  private deserializeField(
    field: TLBField,
    slice: Slice,
    variables: Map<string, number>
  ): any {
    return this.deserializeFieldType(field.fieldType, slice, variables);
  }

  private deserializeFieldType(
    fieldType: TLBFieldType,
    slice: Slice,
    variables: Map<string, number>
  ): any {
    const evaluator = new MathExprEvaluator(variables);

    switch (fieldType.kind) {
      case "TLBNumberType": {
        const bits = evaluator.evaluate(fieldType.bits);
        const value = slice.loadUint(bits);
        return fieldType.signed ? this.toSignedNumber(value, bits) : value;
      }

      case "TLBBoolType": {
        if (fieldType.value !== undefined) {
          return fieldType.value;
        }
        return slice.loadBit() === true;
      }

      case "TLBBitsType": {
        const bits = evaluator.evaluate(fieldType.bits);
        return slice.loadBits(bits);
      }

      case "TLBNamedType": {
        const type = this.tlbCode.types.get(fieldType.name);

        if (fieldType.name === "Bool") {
          return slice.loadBit() === true;
        }

        if (!type) {
          throw new Error(`Type ${fieldType.name} not found in TL-B schema`);
        }

        return this.deserializeType(type, slice, fieldType.arguments);
      }

      case "TLBCoinsType": {
        return slice.loadCoins();
      }

      case "TLBAddressType": {
        return slice.loadAddress();
      }

      case "TLBCellType": {
        return slice.loadRef();
      }

      case "TLBCellInsideType": {
        const ref = slice.loadRef();
        const refSlice = ref.beginParse();
        return this.deserializeFieldType(fieldType.value, refSlice, variables);
      }

      case "TLBHashmapType": {
        const keySize = evaluator.evaluate(fieldType.key.expr);
        const dict = slice.loadDict(Dictionary.Keys.BigInt(keySize), {
          serialize: () => {
            /* NO_USED */
          },
          parse: (slice) =>
            this.deserializeFieldType(
              fieldType.value,
              slice,
              new Map(variables)
            ),
        });
        return dict;
      }

      case "TLBVarIntegerType": {
        const size = evaluator.evaluate(fieldType.n);
        if (fieldType.signed) {
          return slice.loadVarIntBig(size);
        } else {
          return slice.loadVarUintBig(size);
        }
      }

      case "TLBMultipleType": {
        const times = evaluator.evaluate(fieldType.times);
        const result = [];
        for (let i = 0; i < times; i++) {
          result.push(
            this.deserializeFieldType(fieldType.value, slice, variables)
          );
        }
        return result;
      }

      case "TLBCondType": {
        const condition = evaluator.evaluate(fieldType.condition);
        if (condition) {
          return this.deserializeFieldType(fieldType.value, slice, variables);
        }
        return null;
      }

      case "TLBTupleType": {
        const cell = slice.loadRef();
        return parseTuple(cell);
      }

      default:
        throw new Error(`Unsupported field type: ${fieldType.kind}`);
    }
  }

  private serializeType(type: TLBType, data: any, builder: Builder): void {
    // Find matching constructor by kind
    const constructorName = data.kind.substring(type.name.length + 1); // Remove TypeName_ prefix

    const constructor = type.constructors.find(
      (c) => c.name === constructorName
    );
    if (!constructor) {
      throw new Error(
        `Constructor ${constructorName} not found for type ${type.name}`
      );
    }

    // Store tag if present
    if (constructor.tag.bitLen > 0) {
      const tag = parseInt(constructor.tag.binary, 2);
      builder.storeUint(tag, constructor.tag.bitLen);
    }

    // Initialize variables map for constraint evaluation
    const variables = new Map<string, number>();

    // Serialize fields
    for (const field of constructor.fields) {
      if (!field.anonymous) {
        this.serializeField(field, data[field.name], builder, variables);
      } else {
        // For anonymous fields, we need to extract from constraints or use default
        // This is a simplified approach, would need more complex logic for real cases
        this.serializeField(field, null, builder, variables);
      }
    }

    // Check constraints
    const evaluator = new MathExprEvaluator(variables);
    for (const constraint of constructor.constraints) {
      if (evaluator.evaluate(constraint) !== 1) {
        throw new Error(
          `Constraint failed for type ${type.name}, constructor ${constructor.name}`
        );
      }
    }
  }

  private serializeField(
    field: TLBField,
    value: any,
    builder: Builder,
    variables: Map<string, number>
  ): void {
    this.serializeFieldType(field.fieldType, value, builder, variables);
  }

  private serializeFieldType(
    fieldType: TLBFieldType,
    value: any,
    builder: Builder,
    variables: Map<string, number>
  ): void {
    const evaluator = new MathExprEvaluator(variables);

    switch (fieldType.kind) {
      case "TLBNumberType": {
        const bits = evaluator.evaluate(fieldType.bits);
        builder.storeUint(value, bits);
        break;
      }

      case "TLBBoolType": {
        if (fieldType.value !== undefined) {
          // Fixed value, nothing to store
          break;
        }
        builder.storeBit(value ? 1 : 0);
        break;
      }

      case "TLBBitsType": {
        const bits = evaluator.evaluate(fieldType.bits);
        // TODO: check if value is a bitstring
        // builder.storeBits(value, bits);
        builder.storeBits(value);
        break;
      }

      case "TLBNamedType": {
        const type = this.tlbCode.types.get(fieldType.name);
        if (!type) {
          throw new Error(`Type ${fieldType.name} not found in TL-B schema`);
        }
        this.serializeType(type, value, builder);
        break;
      }

      case "TLBCoinsType": {
        builder.storeCoins(value);
        break;
      }

      case "TLBAddressType": {
        builder.storeAddress(value);
        break;
      }

      case "TLBCellType": {
        builder.storeRef(value);
        break;
      }

      case "TLBCellInsideType": {
        const nestedBuilder = beginCell();
        this.serializeFieldType(
          fieldType.value,
          value,
          nestedBuilder,
          variables
        );
        builder.storeRef(nestedBuilder.endCell());
        break;
      }

      case "TLBHashmapType": {
        const keySize = evaluator.evaluate(fieldType.key.expr);
        const dict = Dictionary.empty(
          Dictionary.Keys.BigInt(keySize),
          Dictionary.Values.Cell()
        );

        if (value) {
          for (const [key, dictValue] of Object.entries(value)) {
            const valueBuilder = beginCell();
            this.serializeFieldType(
              fieldType.value,
              dictValue,
              valueBuilder,
              new Map(variables)
            );
            dict.set(BigInt(key), valueBuilder.endCell());
          }
        }

        builder.storeDict(dict);
        break;
      }

      case "TLBVarIntegerType": {
        const size = evaluator.evaluate(fieldType.n);
        if (fieldType.signed) {
          builder.storeVarInt(value, size);
        } else {
          builder.storeVarUint(value, size);
        }
        break;
      }

      case "TLBMultipleType": {
        const times = evaluator.evaluate(fieldType.times);
        for (let i = 0; i < times; i++) {
          this.serializeFieldType(
            fieldType.value,
            value[i],
            builder,
            variables
          );
        }
        break;
      }

      case "TLBCondType": {
        const condition = evaluator.evaluate(fieldType.condition);
        if (condition) {
          this.serializeFieldType(fieldType.value, value, builder, variables);
        }
        break;
      }

      case "TLBTupleType": {
        const cell = serializeTuple(value);
        builder.storeRef(cell);
        break;
      }

      default:
        throw new Error(`Unsupported field type: ${fieldType.kind}`);
    }
  }

  private toSignedNumber(value: number, bits: number): number {
    const maxUnsigned = 1 << bits;
    const signBit = 1 << (bits - 1);

    if (value & signBit) {
      return value - maxUnsigned;
    }

    return value;
  }
}

// Export a simple API for users
export function parseTLB(schema: string): TLBRuntime {
  return TLBRuntime.fromTLB(schema);
}

export { Builder, Slice, Cell, beginCell };
