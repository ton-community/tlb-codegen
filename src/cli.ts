#!/usr/bin/env node
import meow from 'meow';

import { generateCode } from './node';

const cli = meow(
    `
	Usage
	  $ tlb <tlbpath>

	Options
	  --output, -o Output file, default = tlbpath with .ts extension e.g. ./path/to/file.tlb -> ./path/to/file.ts
	  --language, -l Programming language result. Supported languages: ["typescript"]. Default: "typescript"
	  
	Examples
	  $ tlb ./path/to/file.tlb
	  > Generated code is saved to ./path/to/file.ts

	  $ tlb -o ./path/to/output.ts ./path/to/file.tlb
	  > Generated code is saved to ./path/to/output.ts

	  $ tlb -l typescript ./path/to/file.tlb
	  > Generated code is saved to ./path/to/file.ts
`,
    {
        flags: {
            output: {
                type: 'string',
                alias: 'o',
            },
            language: {
                type: 'string',
                alias: 'l',
            },
        },
    },
);

let input = cli.input.at(0);
if (input) {
    if (input.match(/\.tlb$/) == null) {
        // eslint-disable-next-line no-console
        console.error('Input file must have .tlb extension');
        process.exit(1);
    }

    let output = input.replace('.tlb', '.ts');
    if (cli.flags.output) {
        output = cli.flags.output;
    }
    let language = 'typescript';
    if (cli.flags.language) {
        language = cli.flags.language;
    }
    generateCode(input, output, language);
}
