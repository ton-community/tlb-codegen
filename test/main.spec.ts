import { describe, expect, test } from '@jest/globals';

import { generateCode, getTLBCode } from '../src';

describe('main', () => {
    test('generateCode', () => {
        expect(generateCode('_ x:# = Foo;', 'typescript')).toMatchSnapshot();
    });

    test('getTLBCode', () => {
        expect(getTLBCode('_ x:# = Foo;')).toMatchSnapshot();
    });
});
