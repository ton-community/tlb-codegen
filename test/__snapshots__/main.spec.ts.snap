// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main generateCode 1`] = `
"import { Builder } from '@ton/core'
import { Slice } from '@ton/core'
import { beginCell } from '@ton/core'
import { BitString } from '@ton/core'
import { Cell } from '@ton/core'
import { Address } from '@ton/core'
import { ExternalAddress } from '@ton/core'
import { Dictionary } from '@ton/core'
import { DictionaryValue } from '@ton/core'
import { TupleItem } from '@ton/core'
import { parseTuple } from '@ton/core'
import { serializeTuple } from '@ton/core'
export function bitLen(n: number) {
    return n.toString(2).length;
}

export interface Bool {
    readonly kind: 'Bool';
    readonly value: boolean;
}

export function loadBool(slice: Slice): Bool {
    if (slice.remainingBits >= 1) {
        let value = slice.loadUint(1);
        return {
            kind: 'Bool',
            value: value == 1
        }

    }
    throw new Error('Expected one of "BoolFalse" in loading "BoolFalse", but data does not satisfy any constructor');
}

export function storeBool(bool: Bool): (builder: Builder) => void {
    return ((builder: Builder) => {
        builder.storeUint(bool.value ? 1: 0, 1);
    })

}



export function loadBoolFalse(slice: Slice): Bool {
  if (((slice.remainingBits >= 1) && (slice.preloadUint(1) == 0b0))) {
      slice.loadUint(1);
      return {
          kind: 'Bool',
          value: false
      }

  }
  throw new Error('Expected one of "BoolFalse" in loading "BoolFalse", but data does not satisfy any constructor');
}

export function loadBoolTrue(slice: Slice): Bool {
  if (((slice.remainingBits >= 1) && (slice.preloadUint(1) == 0b1))) {
      slice.loadUint(1);
      return {
          kind: 'Bool',
          value: true
      }

  }
  throw new Error('Expected one of "BoolTrue" in loading "BoolTrue", but data does not satisfy any constructor');
}

export function copyCellToBuilder(from: Cell, to: Builder): void {
    let slice = from.beginParse();
    to.storeBits(slice.loadBits(slice.remainingBits));
    while (slice.remainingRefs) {
        to.storeRef(slice.loadRef());
    }
}
// _ x:# = Foo;

export interface Foo {
    readonly kind: 'Foo';
    readonly x: number;
}

// _ x:# = Foo;

export function loadFoo(slice: Slice): Foo {
    let x: number = slice.loadUint(32);
    return {
        kind: 'Foo',
        x: x,
    }

}

export function storeFoo(foo: Foo): (builder: Builder) => void {
    return ((builder: Builder) => {
        builder.storeUint(foo.x, 32);
    })

}

"
`;

exports[`main getTLBCode 1`] = `
TLBCode {
  "types": Map {
    "Foo" => TLBType {
      "constructors": [
        TLBConstructor {
          "constraints": [],
          "declaration": "_ x:# = Foo;",
          "fields": [
            {
              "anonymous": false,
              "fieldType": {
                "bits": TLBNumberExpr {
                  "hasNeg": false,
                  "n": 32,
                  "variables": Set {},
                },
                "kind": "TLBNumberType",
                "maxBits": 32,
                "signed": false,
                "storeBits": TLBNumberExpr {
                  "hasNeg": false,
                  "n": 32,
                  "variables": Set {},
                },
              },
              "name": "x",
              "subFields": [],
            },
          ],
          "name": "_",
          "parameters": [],
          "parametersMap": Map {},
          "tag": {
            "binary": "",
            "bitLen": 0,
          },
          "tlbType": "Foo",
          "variables": [
            TLBVariable {
              "deriveExpr": undefined,
              "initialExpr": undefined,
              "isConst": false,
              "isField": true,
              "name": "x",
              "negated": false,
              "type": "#",
            },
          ],
          "variablesMap": Map {
            "x" => TLBVariable {
              "deriveExpr": undefined,
              "initialExpr": undefined,
              "isConst": false,
              "isField": true,
              "name": "x",
              "negated": false,
              "type": "#",
            },
          },
        },
      ],
      "name": "Foo",
    },
  },
}
`;
